### ISSUE TITLE
Enable node reuse on .NET Core

### ISSUE THREAD

**Issue by AndyGerlicher** [Microsoft/msbuild#2703]
Support `/nodereuse` command to improve inner-loop performance.

**Comment by mmitche** [comment]
@AndyGerlicher This needs to be done with care.  We had severe problems with nodes living from build to build and breaking when build task dlls would change.

**Comment by AndyGerlicher** [comment]
There isn't a plan to change that aspect of it. We'll want to turn it off in CI.

**Comment by danmosemsft** [comment]
@mmitche node reuse is a good fit for the local dev loop where the startup cost matters and they are unlikely to be modifying the task assemblies, and can disable it if needed. CI should definitely disable it.

**Comment by mmitche** [comment]
I think it could work in CI too.  Is there a way to ensure that loaded task assemblies match any new assemblies that are about to be loaded?  That would solve previously seen problems too.

**Comment by davkean** [comment]
Are you tracking/planning on changing how node reuse works @AndyGerlicher so that you can get the benefit of it for a single project?

**Comment by rainersigwald** [comment]
To get this to work, we need:

- [ ] To enable node-finding on core
  - [ ] Can just attempt to connect to every `dotnet` process as a worker node
  - [ ] Ideally would look at command lines to see if processes are worker nodes at all
- [ ] A mechanism to ensure that a user with a low-priv MSBuild can't get connected to by a high-priv MSBuild and send arbitrary build commands. Currently planning to use a new `PipeOptions` setting: https://github.com/dotnet/corefx/issues/25427
- [ ] (Nice but not strictly necessary) Simplify the pipe-handling code to be unified (.NET Core 2 has most of what we needed to ifdef out for pre-1.0)
- [ ] Testing on at least a Linux, macOS, and Windows.

Changing how node reuse works is out of scope for this item.

**Comment by danmosemsft** [comment]
cc @safern who is in progress on https://github.com/dotnet/corefx/issues/25427

**Comment by rainersigwald** [comment]
https://github.com/Microsoft/msbuild/pull/2961 is working with everything except `PipeOptions.CurrentUserOnly`.

**Comment by danmosemsft** [comment]
@rainersigwald I expect @pjanotti to get the `PipeOptions.CurrentUserOnly` PR into master next week.

**Comment by safern** [comment]
@rainersigwald this work is done in the corefx side. @pjanotti merged this ~ last week into master.

**Comment by rainersigwald** [comment]
@jeffkl is picking #2961 up. Currently blocked on adopting the latest corefx build because of behavior changes that broke us in other ways--see #3061.
